File: scripts/producer/guided_caption_profile.py
"""Explicit bounded caption-preset intent, not transcription or visual approval.

The initial caption profile uses the existing captured Producer line/karaoke
defaults for ALL kept words. It does not infer custom typography, corrections,
groups, placement, scene suppression, chapters or legacy subtitle intent.
"""
from __future__ import annotations

from captions.caption_contract import validate_caption_track

CAPTION_PROFILE = "unity-source-float-own-screen-caption-pages-v1"
CAPTION_SHORT_PROFILE = "unity-source-float-manual-short-own-screen-caption-pages-v1"
CAPTION_BODY_PROFILE = "held-source-float-own-screen-caption-pages-body-v1"
CAPTION_SHORT_BODY_PROFILE = "held-source-float-manual-short-own-screen-caption-pages-body-v1"
SCREENED_CAPTION_PROFILE = "unity-source-float-own-screen-caption-layout-v2"
SCREENED_CAPTION_SHORT_PROFILE = "unity-source-float-manual-short-own-screen-caption-layout-v2"
SCREENED_CAPTION_BODY_PROFILE = "held-source-float-own-screen-caption-layout-body-v2"
SCREENED_CAPTION_SHORT_BODY_PROFILE = "held-source-float-manual-short-own-screen-caption-layout-body-v2"
_EXCLUDED = ("captionCorrectionLedger", "captionStyles", "captionChapters", "dialogueCaptionAuthority")


def caption_profile(value: object) -> bool:
    """No old token or omitted value silently opts into caption execution."""
    from guided_presenter_profile import presenter_caption_profile

    return type(value) is str and value in (CAPTION_PROFILE, CAPTION_SHORT_PROFILE,
                                            SCREENED_CAPTION_PROFILE, SCREENED_CAPTION_SHORT_PROFILE) \
        or presenter_caption_profile(value)


def screened_caption_profile(value: object) -> bool:
    """Only explicit new opening tokens require real caption/graphic screening."""
    from guided_presenter_profile import presenter_caption_profile

    return type(value) is str and value in (SCREENED_CAPTION_PROFILE, SCREENED_CAPTION_SHORT_PROFILE) \
        or presenter_caption_profile(value)


def caption_screen_fields(profile: object) -> set[str]:
    """Old results cannot acquire a new-policy screening statement."""
    from guided_presenter_profile import PRESENTER_CAPTION_BODY_PROFILE, PRESENTER_CAPTION_SHORT_BODY_PROFILE

    return {"captionLayoutScreen"} if screened_caption_profile(profile) or profile in (
        SCREENED_CAPTION_BODY_PROFILE, SCREENED_CAPTION_SHORT_BODY_PROFILE,
        PRESENTER_CAPTION_BODY_PROFILE, PRESENTER_CAPTION_SHORT_BODY_PROFILE) else set()


def caption_preset_plan(plan: dict) -> dict:
    """Validate the exact submitted preset without normalizing the candidate."""
    target = plan.get("target")
    if type(target) is not dict or target.get("mode") not in ("short", "longform"):
        raise RuntimeError("caption preset needs an explicit supported target")
    expected = (1080, 1920) if target["mode"] == "short" else (1920, 1080)
    if any(type(target.get(key)) is not int for key in ("width", "height")) \
            or (target["width"], target["height"]) != expected:
        raise RuntimeError("caption preset requires the existing exact native compiler destination")
    captions = plan.get("captions")
    if type(captions) is not dict or set(captions) != {"burn"} or captions["burn"] is not True:
        raise RuntimeError("caption preset requires only explicit captions.burn=true")
    track = plan.get("captionsTrack")
    if type(track) is not dict or set(track) != {"schemaVersion", "source", "defaultPolicy", "groups"} \
            or type(track.get("schemaVersion")) is not int or track["schemaVersion"] != 1 \
            or track.get("defaultPolicy") not in ("line", "karaoke") or track.get("groups") != []:
        raise RuntimeError("caption preset requires explicit all-kept line/karaoke with no custom groups")
    validated = validate_caption_track(track)
    if validated != track:
        raise RuntimeError("caption preset must not normalize submitted caption authority")
    if any(key in plan for key in _EXCLUDED) or plan.get("chapters"):
        raise RuntimeError("caption preset has unqualified correction/style/chapter/alternate authority")
    if any(plan.get(key) for key in ("overlays", "presenter")):
        raise RuntimeError("caption preset has unsupported additional visual intent")
    return track


def caption_full_fields(profile: object) -> set[str]:
    """Historical full-program records keep their exact original field set."""
    return {"captionProjection"} if caption_profile(profile) else set()

File: src/lib/producer/contracts/guided-caption-profile.ts
/** Explicit initial all-kept caption presets; not typography, ASR or execution approval. */
import { exactKeys, objectValue } from "./validation";
import { parsePositiveRationalV1 } from "./positive-rational";

export const CAPTION_PROFILE = "unity-source-float-own-screen-caption-pages-v1" as const;
export const CAPTION_SHORT_PROFILE = "unity-source-float-manual-short-own-screen-caption-pages-v1" as const;
export const CAPTION_BODY_PROFILE = "held-source-float-own-screen-caption-pages-body-v1" as const;
export const CAPTION_SHORT_BODY_PROFILE = "held-source-float-manual-short-own-screen-caption-pages-body-v1" as const;
export const SCREENED_CAPTION_PROFILE = "unity-source-float-own-screen-caption-layout-v2" as const;
export const SCREENED_CAPTION_SHORT_PROFILE = "unity-source-float-manual-short-own-screen-caption-layout-v2" as const;
export const SCREENED_CAPTION_BODY_PROFILE = "held-source-float-own-screen-caption-layout-body-v2" as const;
export const SCREENED_CAPTION_SHORT_BODY_PROFILE = "held-source-float-manual-short-own-screen-caption-layout-body-v2" as const;

/** An execution requirement, never proof that the actual layout passed. */
export function isScreenedCaptionProfile(value: unknown): boolean {
  return value === SCREENED_CAPTION_PROFILE || value === SCREENED_CAPTION_SHORT_PROFILE;
}

export function isCaptionProfile(value: unknown): boolean {
  return value === CAPTION_PROFILE || value === CAPTION_SHORT_PROFILE || isScreenedCaptionProfile(value);
}

/** No defaults/mutation or ignored custom/group/correction fields. Python independently runs the real compiler. */
export function assertCaptionPresetPlan(plan: Record<string, unknown>): void {
  const target = objectValue(plan.target, "caption target");
  const size = target.mode === "short" ? [1080, 1920] : target.mode === "longform" ? [1920, 1080] : [];
  if (!size.length || target.width !== size[0] || target.height !== size[1]) throw new Error("Caption preset requires exact native compiler destination");
  const captions = objectValue(plan.captions, "caption burn");
  exactKeys(captions, ["burn"], ["burn"], "caption burn");
  if (captions.burn !== true) throw new Error("Caption preset requires explicit captions.burn=true");
  const track = objectValue(plan.captionsTrack, "caption track"), keys = ["schemaVersion", "source", "defaultPolicy", "groups"];
  exactKeys(track, keys, keys, "caption preset track");
  if (track.schemaVersion !== 1 || track.source !== "kept-transcript" || (track.defaultPolicy !== "line" && track.defaultPolicy !== "karaoke")
      || !Array.isArray(track.groups) || track.groups.length) throw new Error("Caption preset requires all-kept line/karaoke without custom groups");
  if (["captionCorrectionLedger", "captionStyles", "captionChapters", "dialogueCaptionAuthority"].some((key) => Object.hasOwn(plan, key))
      || [plan.chapters, plan.overlays, plan.presenter].some(active)) throw new Error("Caption preset has unqualified correction/style/chapter/visual authority");
}

function active(value: unknown): boolean {
  if (Array.isArray(value)) return value.length > 0;
  if (value && typeof value === "object") return Object.keys(value).length > 0;
  return Boolean(value);
}

/** Conservative entire-program page upper bound, not only simultaneous captions or opening pages. */
export function assertCaptionGraphWorkload(plan: Record<string, unknown>, clock: { frameRate: string; totalFrames: number }, graphicCount: number): void {
  const target = objectValue(plan.target, "caption canvas");
  const [numerator, denominator = "1", extra] = clock.frameRate.split("/");
  if (extra !== undefined) throw new Error("Caption workload frame rate is malformed");
  const rate = parsePositiveRationalV1({ numerator, denominator });
  if (!Number.isSafeInteger(clock.totalFrames) || clock.totalFrames <= 0 || !Number.isSafeInteger(graphicCount) || graphicCount < 0) {
    throw new Error("Caption workload needs exact positive frame/count metadata");
  }
  const pageFrames = Math.max(1, Number(BigInt(rate.numerator) * BigInt(30) / BigInt(rate.denominator)));
  const pages = Math.ceil(clock.totalFrames / pageFrames), count = pages + graphicCount;
  const pixels = Number(target.width) * Number(target.height);
  if (!Number.isSafeInteger(pixels) || pixels <= 0 || count > 128 || (count + 1) * pixels > 64 * 1024 * 1024) {
    throw new Error("Caption combined graph exceeds existing64MiPixel workload; no inputs may be dropped");
  }
}
